main.go

```diff
@@ -146,6 +147,12 @@ func main() {
 		setupLog.Error(err, "unable to create controller", "controller", "TransportURL")
 		os.Exit(1)
 	}
+	if strings.ToLower(os.Getenv("ENABLE_WEBHOOKS")) != "false" {
+		if err = (&corev1beta1.OpenStackControlPlane{}).SetupWebhookWithManager(mgr); err != nil {
+			setupLog.Error(err, "unable to create webhook", "webhook", "OpenStackControlPlane")
+			os.Exit(1)
+		}
+	}
```

Comment:
Just a heads-up regarding webhooks and running the operator locally.  There is a scenario where webhooks can complicate dev envs.  Sometimes for dev purposes we want to install the operator via OLM and then remove the actual `openstack-operator-controller-manager` deployment from the CSV, but leave the other children controller-managers (i.e Glance, Cinder, Keystone, etc) running while we iterate on the `openstack-operator-controller-manager` and use `GOWORK= OPERATOR_TEMPLATES=./templates make run` to run the manager locally.  In such a case, the `ValidatingWebhookConfiguration` installed by OLM will cause errors to occur whenever you try to create/update `OpenStackControlPlane` CRs, like so:

```
oc kustomize /home/abays/rh/install_yamls/out/openstack/openstack/cr | oc apply -f -
Error from server (InternalError): error when creating "STDIN": Internal error occurred: failed calling webhook "vopenstackcontrolplane.kb.io": failed to call webhook: Post "https://openstack-operator-controller-manager-service.openstack.svc:443/validate-core-openstack-org-v1beta1-openstackcontrolplane?timeout=10s": dial tcp 10.217.0.102:9443: connect: connection refused
```

This is because the webhook server isn't running in the local process (as per `ENABLE_WEBHOOKS=false`).  Leaving the `ENABLE_WEBHOOKS=false` env var out of course results in a different error because the cert for the web server can't be found outside of an OLM deployment (but there might be a way to work around that).

Anyhow, I just wanted to mention this.  We need to start using webhooks though, so it's just something we'll have to deal with going forward.

Reply:
I suppose another solution is to perhaps patch the CSV to remove the webhook definition.

Reply:
@abays Thanks for checking the PR. I've not been able to get the webhook work either in CI/locally(crc). It failed in CI[1] with the same error. I deployed cert-manager operator locally (missing in CI I guess) and could also see webhook configuration is fine[2], but it seems apiserver can't talk to openstack-operator-controller-manager(webhook) for some reason.

It looks like some routing issue rather than cert validation. Not sure what I'm missing. Do we test webhooks with osp-director-operator?

[1] https://gcsweb-ci.apps.ci.l2s4.p1.openshiftapps.com/gcs/origin-ci-test/pr-logs/pull/openstack-k8s-operators_openstack-operator/155/pull-ci-openstack-k8s-operators-openstack-operator-master-openstack-operator-build-deploy/1617716507368755200/build-log.txt

[2] 

$ oc get ValidatingWebhookConfiguration/vopenstackcontrolplane.kb.io-65ffw -o yaml
....
webhooks:
- admissionReviewVersions:
  - v1
  clientConfig:
    caBundle: <removed>
    service:
      name: openstack-operator-controller-manager-service
      namespace: openstack
      path: /validate-core-openstack-org-v1beta1-openstackcontrolplane
      port: 443

$oc get svc/openstack-operator-webhook-service -o yaml
....
spec:
  clusterIP: 10.217.5.30
  clusterIPs:
  - 10.217.5.30
  internalTrafficPolicy: Cluster
  ipFamilies:
  - IPv4
  ipFamilyPolicy: SingleStack
  ports:
  - port: 443
    protocol: TCP
    targetPort: 9443
  selector:
    control-plane: controller-manager
  sessionAffinity: None
  type: ClusterIP

Reply:
Is the webhook working for you when you deploy via OLM, however?  Or is it not working ever at all?

Reply:
It's not working with OLM bundles too. I linked the CI openstack-operator-build-deploy job where it failed with the same error. I just worked around in CI with the latest patchset by setting failurePolicy: Ingore, as it's just scaffolding atm.

Reply:
Ah okay.  We definitely have webhooks working in OSPdO if you want to consult that as a reference, but the state of the operator-SDK support for webhooks was different back then (so what we have there is not necessarily the "right" way to do it anymore).  I did make a pass at a mutating webhook for the Glance operator a while back using the new operator-SDK approach and it was working for me: https://github.com/openstack-k8s-operators/glance-operator/pull/52.  Even though it's a mutating webhook, the basic infrastructure is mostly the same as validating webhooks.  Maybe something in that PR can help provide clues.  I can also try your PR via OLM and see if I can find the problem.

Reply:
I checked the glance-operator PR and did not find any notable difference. Probably the operator-sdk is doing something wrong which I'm not able to trace. The error is weird to decipher. Initially cert-manager operator was missing on my setup and I did deploy that. Maybe some iptable rule missing or incorrectly configured on the nodes. If you can try the PR and find the issue that would be great:)

Reply:
So the issue was "selector labels" for the controller manager/webhook services. It's the same for all operators and the calls are routed to controller pods that don't have the webhook running (load-balanced across all endpoints with the same lebel). It worked at times when it would end up with the openstack-operator-controller. After changing the lables for openstack-operator-controller in this PR, webhook works fine.